### PR TITLE
Fix line folding

### LINKED ISSUE #41: whitespace newline whitespace in folded string causes errors in `ScalarStyle.FOLDED`
**Example**

```dart
import 'package:yaml/yaml.dart';
import 'package:yaml_edit/yaml_edit.dart';

void main() {
  final doc = YamlEditor('''
"hello"
''');

  final node = wrapAsYamlNode(
    'hello \n world',
    scalarStyle: ScalarStyle.FOLDED,
  );
  doc.update([], node);

  print(doc);
}
```

We need to:
 * fix this error,
 * write a test cases to ensure it doesn't happen again.

Maybe, there are some cases where we have to output as `"hello \n world"` in double quoted mode. If we can't find a reasonably safe way to output in _folded_ style, then it's safer to fallback to double quoted mode.

But ideally, we won't have to fallback in all non-trivial cases.

### LINKED ISSUE #44: Code coverage for `>+`
We have no test that exercise this code:
https://coveralls.io/builds/57565406/source?filename=lib%2Fsrc%2Fstrings.dart#L103

That might be a good idea to have some tests for. Maybe there is bugs?

### LINKED ISSUE #86: Default chomping indicator in `ScalarStyle.LITERAL`
It seems `ScalarStyle.LITERAL` applies the `stripping` chomping indicator `-` by default. 

https://github.com/dart-lang/yaml_edit/blob/561309eeb5fda2564dd08bbe790c18113dec8aa7/lib/src/strings.dart#L117-L129

`ScalarStyle.FOLDED` at least checks if there is any line break.

https://github.com/dart-lang/yaml_edit/blob/561309eeb5fda2564dd08bbe790c18113dec8aa7/lib/src/strings.dart#L92-L115

## Expected Behaviour
As indicated in the [spec], we should check for both `line breaks` & `trailing spaces` in order to apply the correct chomping indicator for block styles:

- `-` for stripping any trailing line breaks/spaces (strip)
- `+` for preserving trailing line breaks/spaces (keep)
- Omitting them preserves the line breaks but removes the trailing space (clip)

Ideally, we should end up using the `keep` chomping indicator `+` only since we may have no use for the `strip` chomping indicator `-`.

Fixing this may have desirable ripple effects for #41 & #44.

[spec]: https://yaml.org/spec/1.2.2/#812-literal-style:~:text=the%20indicated%20level.-,8.1.1.2.%20Block%20Chomping%20Indicator,content.%20These%20additional%20lines%20are%20not%20subject%20to%20folding.,-The%20chomping%20method

### PR DESCRIPTION
**Pull request by kekavc24**
This PR:
- Improved line folding. Fixes #41
- Closes #44
- Closes #86
---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by kekavc24** [comment]
@jonasfj This is the draft. Forgive the commits from #84, I will rebase this branch once it is merged. 

Also, I added a funky method to reduce the code duplication after several iterations for both `_tryYamlEncodeLiteral` and `_tryYamlEncodeFolded`. Should I just duplicate the code? As both methods attempt to `chomp` strings the same way.

Will add tests for this then.

File: test/string_test.dart [hunk 1]
```diff
@@ -9,6 +9,7 @@ import 'package:yaml_edit/yaml_edit.dart';
 final _testStrings = [
   "this is a fairly' long string with\nline breaks",
   'whitespace\n after line breaks',
+  'whitespace\n \nbetween line breaks',
```
Comment (jonasfj) [inline test/string_test.dart:12]: by the way, feel free to add lots of cases here.

I'd honestly suggest adding anything we think could cause a problem. 
```suggestion
  'whitespace\n \nbetween line breaks',
  'whitespace and line breaks at end 1\n ',
  'whitespace and line breaks at end 2 \n \n',
  'whitespace and line breaks at end 3 \n\n',
  'whitespace and line breaks at end 3 \n\n ',
```
If we find something that's still broken, we can just comment it out, and file a bug to fix it later.

File: lib/src/strings.dart [hunk 2]
```diff
@@ -94,38 +95,103 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
+      string.endsWith('\n') || string.endsWith(' ') ? '+' : '-';
+
+  final indent = ' ' * indentSize;
+
+  return (
+    '$blockIndicator$chompingIndicator$lineEnding',
+    indent,
+
+    /// Assumes the user did not try to account for windows documents by using
+    /// `\r\n` already
```
Comment (jonasfj) [inline lib/src/strings.dart:191]: We don't have to state that. The user should never account for encoding details.

Sematically, a YAML document consists of YamlNodes, see package:yaml
If a YamlNode is a YamlScalar then it can have a `YamlScalar.value` that is a string.
If `value` is a string, then that is the string that must be returned when the document is parsed.

If the user actually gives us a string that contains `\r\n` we can't encode it in folded or literal mode. The only way to represent a `\r` character is to do quoted string `"\r\n"` -- am I wrong? Is it possible to string escaping in folded or literal strings? I imagine it's not possible, but I could be wrong.

File: lib/src/strings.dart [hunk 3]
```diff
@@ -94,38 +95,103 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
+}
 
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
+/// Prepares a string to be encoded either via [_tryYamlEncodeLiteral] or
+/// [_tryYamlEncodeFolded] when in [CollectionStyle.BLOCK].
+(
+  String literalPrefix,
+  String indent,
```
Comment (jonasfj) [inline lib/src/strings.dart:173]: Returning this seems a bit extreme. The caller already has `indentSize`, why can't the caller just do `' ' * indentSize`, I think should be fine.

In general, I'd suggest we only pass around indentation as an integer.

File: lib/src/strings.dart [hunk 4]
```diff
@@ -94,38 +95,103 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
+  String trimmedString,
+  String strippedString,
+) _prepareBlockString(
+    String blockIndicator, String string, int indentSize, String lineEnding) {
```
Comment (jonasfj) [inline lib/src/strings.dart:177]: btw, feel free to insert assertions, they are practically free when a program is compiled in release mode. And a few sanity checks like this can only help catch bugs :D


```suggestion
    String blockIndicator, String string, int indentSize, String lineEnding) {
    assert(indentSize >= 0);
```

File: lib/src/strings.dart (hunk 2 above)
Comment (jonasfj) [reply lib/src/strings.dart:191]: I'd suggest you take a look at:
https://github.com/dart-lang/yaml_edit/blob/main/lib/src/editor.dart#L564-L583

Essentially, if I use `YamlEditor.update(path, foo)` then `deepEquals(YamlEditor.parseAt(path), foo) == true`. And this MUST always be true (assuming that `deepEquals` ignores the styling and only looks at `.value`).
No matter what string I insert or what style I attempt to use, YamlEditor must fallback to use a less pretty style, if it can't represent a string in the style I chose.

For example, if I try to use special characters or `\0` null bytes or whatever, then this should still hold. But YamlEditor might have to fallback to quoted strings, even if I requested a Folded string style. The style is just a hint, not a requirement, because not all styles can be used for all values.

File: lib/src/strings.dart (hunk 4 above)
Comment (kekavc24) [reply lib/src/strings.dart]: I realised there is no need to `assert`. We can pass in the `indent` and just apply to all strings that appear after `\n`

File: lib/src/strings.dart [hunk 5]
```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeFolded(String string, int indentation, String lineEnding) {
-  String result;
-
-  final trimmedString = string.trimRight();
-  final removedPortion = string.substring(trimmedString.length);
+String _tryYamlEncodeFolded(String string, int indentSize, String lineEnding) {
```
Comment (jonasfj) [inline lib/src/strings.dart:98]: This might be bigger than what you set out to do. But reading the code, I thinking that maybe we should restructure it a bit.


------------------

Maybe we should change how `_tryYamlEncodeFolded` and `_tryYamlEncodeLiteral` works.

What if we change the signature to:
```dart
/// Try to encode [value] as a _YAML folded string_ indent with [indent] spaces.
///
/// This method will apply an appropriate _chomping indicator_, if this is necessary
/// in order to correctly encode [value].
///
/// **Example** of a folded string.
/// ```yaml
/// key: >-
///   my folded
///   string
/// ```
///
/// See: https://yaml.org/spec/1.2.2/#813-folded-style
///
/// Returns `null`, if [value] cannot be encoded as a YAML folded string.
String? _tryYamlEncodeFolded(String value, int indent, String lineEnding) 
```

So we return `null`, when the value can't be encoded correctly as a folded string.

Then we get to change `yamlEncodeBlockScalar` such that it tries to encode in the style hinted by `YamlScalar.style`, and if this fails, we can fallback to a different block-mode, and if all block-modes fail we can fallback to quoted string mode.

This way, we don't have to write `yamlEncodeBlockScalar` such that it figures out of the value can be encoded in folded mode, and then `_tryYamlEncodeFolded` does the encoding.
Which is also weird, because `_tryYamlEncodeFolded` starts with `try`, but it actually doesn't have a mechanism for communicating back to say "oh, I can't correctly encode this value".

This way: `_tryYamlEncodeFolded` is responsible for:
 * (A) Determining if `value` can be encoded as a folded string.
 * (B) Actually encoding `value` as a folded string.

And `yamlEncodeBlockScalar` is responsible for:
 * (1) Trying `_tryYamlEncodeFolded` or `_tryYamlEncodeLiteral` (depending on `YamlScalar.style`).
 * (2) Falling back to another style if the preferred style couldn't be used.
 * (3) Falling back to using quoted string encoding (which always works), if none of the block styles could be used.

And `yamlEncodeBlockScalar` doesn't have to know what values can be encoded in what styles, it just determines the preferred option and tries it.

**Comment by jonasfj** [comment]
I landed #84 so please rebase this PR :D

**Comment by kekavc24** [comment]
@jonasfj I've updated the change log with changes from #84 too.

File: lib/src/strings.dart (hunk 5 above)
Comment (kekavc24) [reply lib/src/strings.dart:98]: I didn't see this. Lemme get to it.

Comment (jonasfj) [reply lib/src/strings.dart:98]: No rush :)

We can also do it in a follow up.

Btw, there values we can't encode right?
Like special characters and some weird combination of whitespace and line breaks I would also assume.

Comment (kekavc24) [reply lib/src/strings.dart:98]: If so, this PR is good to go. 

> No rush :)
> 
> We can also do it in a follow up.
> 
> Btw, there values we can't encode right? Like special characters and some weird combination of white space and line breaks I would also assume.

Your suggestions in the previous comment are in the ball park of what I had in mind. 

I realised the code below makes the tests pass and we never get to encode the folded string as `>+ my string` as it trims the space as the end including `\n`.

https://github.com/kekavc24/yaml_edit/blob/08a146ef8f2c7aba908d2017c9ec840b882c9e51/lib/src/strings.dart#L186-L196

> No rush :)
> 
> We can also do it in a follow up.
> 
> Btw, there values we can't encode right? Like special characters and some weird combination of white space and line breaks I would also assume.

Yeah. The trailing line breaks and white-space stand out. Your suggestions actually make it possible to correctly debug what the issue is. I'll look into it.

However, I can still tackle it in this PR. No issue. Your call.

File: lib/src/strings.dart [hunk 6]
```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
+/// Prepares a string to be encoded either via [_tryYamlEncodeLiteral] or
+/// [_tryYamlEncodeFolded] when in [CollectionStyle.BLOCK].
+(
+  String literalPrefix,
+  String trimmedString,
+  String strippedString
+) _prepareBlockString(
```
Comment (jonasfj) [inline lib/src/strings.dart:179]: Are we sure this method is worth the dedupliation?

It seems weird that we do `trimmed.replaceAll('\n', lineEnding + indent)` and then inside `_tryYamlEncodeFolded` we split it on `lineEnding`.

It's unclear what we return from this method too.

I think may we should just let this be duplicated into the two method.

Then `_tryYamlEncodeFolded` can split on `\n` instead of splitting on `lineEnding`, which is probably going to result in the same thing, but hopefully going to be less confusing.

File: lib/src/strings.dart [hunk 7]
```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
+
+    // Apply the left indent. First line didn't have indent.
+    final first = indent + trimmedSplit[0];
+
+    if (trimmedSplit.length > 1) {
+      /// Try folding to match specification:
+      /// * https://yaml.org/spec/1.2.2/#65-line-folding
+      trimmedString =
+          trimmedSplit.foldIndexed(first, (index, previous, current) {
+        if (index == 0) return previous;
```
Comment (jonasfj) [inline lib/src/strings.dart:120]: Take a look at `reduceIndexed`, in which case you might not need the index and you can just use `reduce`.

> Calls combine for each element except the first.

Then you probably won't need the ` if (trimmedSplit.length > 1) {... else ....` either.

I suspect that `[foo].reduce(...) == foo` (a list of length one always reduces to the first element).

File: lib/src/strings.dart (hunk 5 above)
Comment (jonasfj) [reply lib/src/strings.dart:98]: > Your suggestions actually make it possible to correctly debug what the issue is. I'll look into it.
>
> I can still tackle it in this PR. No issue. Your call.

It's probably a good idea to spend some time making sure we get this right :D

It'd be really nice if we didn't have so many weird string encoding bugs.

File: lib/src/strings.dart [hunk 8]
```diff
@@ -94,38 +95,100 @@ String _tryYamlEncodeSingleQuoted(String string) {
...
+
+        var updated = current;
+
+        /// For consecutive non-empty lines, we add a `\n` with a space.
+        ///
+        /// We need to ensure the ACTUAL previous string was non-empty
+        if (trimmedSplit[index - 1].trim().isNotEmpty) {
+          /// We remove indent and check if the string starts with a space.
+          ///
+          /// Apply `\n` for `foo\nbar` but not `foo\n bar`.
```
Comment (jonasfj) [inline lib/src/strings.dart:136]: I'm not sure follow 100%, so you're trying to:
 * Insert a line break into `foo\nbar` -> `foo\n\nbar`,
 * But not do that for `foo\n bar`.

With the added complexity that we have indentation, so it's actually `${indent}foo\n${indent}bar` and not `foo\nbar`.

Might I suggest that duplicating the logic in `_prepareBlockString` could make this simpler.
What if we just did `string.trimRigh().split('\n').reduce(...` or something like it.

It probably makes sense to do this along with the change where `_tryYamlEncodeFolded` can return `null`, if the value can't be encoded as a folded string.

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,13 @@
   list.
   ([#69](https://github.com/dart-lang/yaml_edit/issues/69))
 
+- Fix error thrown when inserting in nested list using `spliceList` method
+  ([#83](https://github.com/dart-lang/yaml_edit/issues/83))
```
Comment (jonasfj) [inline CHANGELOG.md:9]: Feel free to file this as a separate PR, if you want. Then we get to land it quickly :D

Tagging it along here is harmless though.

File: lib/src/strings.dart (hunk 6 above)
Comment (kekavc24) [reply lib/src/strings.dart:179]: That was part of the existing code.

> It seems weird that we do `trimmed.replaceAll('\n', lineEnding + indent)` and then inside `_tryYamlEncodeFolded` we split it on `lineEnding`.
> 
> It's unclear what we return from this method too.
> 
> I think may we should just let this be duplicated into the two method.
> 
> Then `_tryYamlEncodeFolded` can split on `\n` instead of splitting on `lineEnding`, which is probably going to result in the same thing, but hopefully going to be less confusing.

I am getting confused. Help me clarify something. If I do,

```dart
final multiLine = '''
My multiline string
is great
''';
```

Will `Dart` return the `line breaks` as `\r\n` when on Windows or just plain `\n`.

File: lib/src/strings.dart (hunk 8 above)
Comment (kekavc24) [reply lib/src/strings.dart:136]: Yeah. `Yaml` uses the first line for indent information. So for `foo\n bar`:
1. Encode as `>-${indent}foo\n${indent} bar`.
2. When parsed, the indent on `foo` allows us to have the space before `bar`

For `foo\nbar`:
1. Encode as `>-${indent}foo\n\n${indent}bar`.
2. When parsed, [yaml] gets rid of the first line break since the next one is also an empty line that giving us `foo\nbar`

[yaml]: https://yaml.org/spec/1.2.2/#65-line-folding:~:text=Line%20folding%20allows%20long%20lines%20to%20be%20broken%20for%20readability%2C%20while%20retaining%20the%20semantics%20of%20the%20original%20long%20line.%20If%20a%20line%20break%20is%20followed%20by%20an%20empty%20line%2C%20it%20is%20trimmed%3B%20the%20first%20line%20break%20is%20discarded%20and%20the%20rest%20are%20retained%20as%20content.

File: lib/src/strings.dart [hunk 10]
```diff
@@ -175,24 +232,19 @@ String yamlEncodeBlockScalar(
...
+          encoded = _tryYamlEncodeFolded(val, indentation, lineEnding);
+        } else if (value.style == ScalarStyle.LITERAL) {
+          encoded = _tryYamlEncodeLiteral(val, indentation, lineEnding);
         }
       }
+
+      if (encoded != null) return encoded;
     }
 
     return _tryYamlEncodePlain(value.value);
```
Comment (jonasfj) [inline lib/src/strings.dart:250]: How about something like this?

We make all the `_try...` methods return `null` if they can't encode correctly.
And they we try the best option first, and just try a few relevant fallbacks.

It's probably rare that literal works when folded doesn't, and vice versa, but it can happen, right? And trying is probably rather cheap.

I realize that we do end up checking `_hasUnprintableCharacters` inside each method. But I'm guess it's fairly cheap, and we'll only run it more than once in the cases where there actually is something unprintable. So it won't really affect performance much (after all, it's rare that there are unprintable characters).

```suggestion
      switch (value.style) {
        // Prefer 'plain', fallback to "double quoted"
        case ScalarStyle.PLAIN:
          return  _tryYamlEncodePlain(value.value) ?? _yamlEncodeDoubleQuoted(val);
        
        // Prefer 'single quoted', fallback to "double quoted"
        case ScalarStyle.SINGLE_QUOTED:
          return _tryYamlEncodeSingleQuoted(val) ?? _yamlEncodeDoubleQuoted(val);

        // Prefer folded string, try literal as fallback
        // otherwise fallback to "double quoted"
        case ScalarStyle.FOLDED:
          return _tryYamlEncodeFolded(val, indentation, lineEnding)
            ?? _tryYamlEncodeLiteral(val, indentation, lineEnding)
            ?? _yamlEncodeDoubleQuoted(val);

        // Prefer literal string, try folded as fallback
        // otherwise fallback to "double quoted"
        case ScalarStyle.LITERAL:
          return _tryYamlEncodeLiteral(val, indentation, lineEnding)
            ?? _tryYamlEncodeFolded(val, indentation, lineEnding)
            ?? _yamlEncodeDoubleQuoted(val);

        // Prefer plain, try 'single quote', fallback to "double quoted"
        default:
          return _tryYamlEncodePlain(value.value)
            ?? _tryYamlEncodeSingleQuoted(val)
            ?? _yamlEncodeDoubleQuoted(val);
      }
    }
```

Comment (kekavc24) [reply lib/src/strings.dart:250]: `_tryYamlEncodePlain` internally falls back to `_yamlEncodeDoubleQuoted`. Should I make it return `null` if it fails instead?

Comment (jonasfj) [reply lib/src/strings.dart:250]: Yeah, let's, then it'll be consistent.

If we do a switch statement like above, we get to deal with one thing in each method. And the heuristics of what we fall back to is all located in the switch statement.

File: lib/src/strings.dart (hunk 6 above)
Comment (jonasfj) [reply lib/src/strings.dart]: I actually don't know, I would suspect it always uses `\n`, because anything just seems crazy :rofl: 

But as far as yaml_edit is concerned I don't see how that's relevant.

If `package:yaml` parses a YAML document that uses `\n\r` as line breaks, it won't included that kind of line breaks in folded or literal strings.

The only way to get that into a string in YAML is to use a double quoted string with escape characters: `"\n\r"`.

I'm like 99% sure that parsing YAML on Windows won't cause multi-line strings to be broken by `\n\r`, they'll always be broken by `\n` once parsed. The actual YAML file on disk should contain `\n\r`, that's correct. But when parsed the `YamlScalar.value` should never be a string that contains `\n\r`, unless it was a double quoted string with an escape sequence.

Does that makes sense? Yes it's easy to get lost, and I'll admit there is a fair 12.3% risk I missed something here :rofl:

File: test/utils_test.dart [hunk 11]
```diff
@@ -318,13 +318,35 @@ strings:
...
+      test('generates literal string with trailing line break and space', () {
+        final doc = YamlEditor('');
+
+        doc.update(
+          [],
+          wrapAsYamlNode('test\ntest\n ', scalarStyle: ScalarStyle.LITERAL),
+        );
+
+        expect(doc.toString(), equals('|+\n  test\n  test\n   '));
+      });
```
Comment (jonasfj) [inline test/utils_test.dart:348]: So I tried something like this in an online editor that converted yaml to JSON and I kept getting the resulting string as: `"test\ntest\n \n"`

Probably because the editor always ended the YAML document with a line-break: `|+\n  test\n  test\n   ` would become: `|+\n  test\n  test\n   \n` (because files normally end with a line break).

I tried to recreate it, and it wasn't a problem. But I'm guessing it's unwise for us to produce a YAML document where adding empty lines to the YAML document causes it to change its semantics.

If I parse the YAML document `|+\n  test\n  test\n   \n`, then I won't get `"test\ntest\n \n"`.

Example:
 * YAML: `A: |+\n    test\n    test\n     `
   * YAML to JSON: `{"A":"test\ntest\n "}` (this is correct!)
 * If I add `\n` at the end of the YAML I get: `A: |+\n    test\n    test\n     \n`
   * YAML to JSON: `{"A":"test\ntest\n \n"}` (this is wrong)
 * If I add `\nB: true` at the end of the YAML I get: `A: |+\n    test\n    test\n     \nB: true`
  * YAML to JSON: `{"A":"test\ntest\n \n","B":true}` (this is wrong)

So if we allow this as a literal string, we are VERY likely to cause all sorts of breakages next time we try to append a key.

Comment (jonasfj) [reply test/utils_test.dart:348]: I did another variant of this test case:
```dart
      test('generates literal string with trailing line break and space (2)',
          () {
        final doc = YamlEditor('key1: true');

        doc.update(
          ['key2'],
          wrapAsYamlNode('test\ntest\n ', scalarStyle: ScalarStyle.LITERAL),
        );
      });
```
it causes a crash:
```
  Assertion failed: (package:yaml_edit) Modification did not result in expected result.
  
  # YAML before edit:
  > key1: true
  
  # YAML after edit:
  > key1: true
  > key2: |+
  >     test
  >     test
  >      
  > 
```

------------------------

I'm beginning to see how a literal string ending in `\n ` might have to be encoded as a double quoted string. :rofl:

File: lib/src/strings.dart [hunk 12]
```diff
@@ -67,142 +56,281 @@ String _yamlEncodeDoubleQuoted(String string) {
...
+    case ScalarStyle.PLAIN:
+      return _tryYamlEncodePlain(value) ?? _yamlEncodeDoubleQuoted(value);
+
+    // Prefer 'single quoted', fallback to "double quoted"
+    case ScalarStyle.SINGLE_QUOTED:
+      return _tryYamlEncodeSingleQuoted(value) ??
+          _yamlEncodeDoubleQuoted(value);
+
+    /// Prefer folded string, try literal as fallback
+    /// otherwise fallback to "double quoted"
```
Comment (jonasfj) [inline lib/src/strings.dart:319]:
```suggestion
    /// Prefer folded string, fallback to "double quoted"
```

File: lib/src/strings.dart [hunk 13]
```diff
@@ -67,142 +56,281 @@ String _yamlEncodeDoubleQuoted(String string) {
...
+    case ScalarStyle.FOLDED:
+      return _tryYamlEncodeFolded(value, indentation, lineEnding) ??
+          _yamlEncodeDoubleQuoted(value);
+
+    /// Prefer literal string, try folded as fallback
+    /// otherwise fallback to "double quoted"
```
Comment (jonasfj) [inline lib/src/strings.dart:325]:
```suggestion
    /// Prefer literal string, fallback to "double quoted"
```

File: test/utils_test.dart (hunk 11 above)
Comment (jonasfj) [reply test/utils_test.dart:348]: Should we consider falling back to double quoted mode for literal strings ending in newline + whitespace?

The whole adding a new key afterwards seems like a reasonable thing, and it's a little sad that this can't work.



Of course, we can't know what document the user gives us. So we could also view this as an orthogonal bug to do with adding keys to block maps and entries to block lists, because those are the scenarios where this could go wrong.

I don't know if there is a reasonably way to fix adding keys to block maps that have a preceeding key-value pair where the value is a literal string ending in `\n ` :rofl:  -- definitely we should not try to fix that in this PR.

Comment (jonasfj) [reply test/utils_test.dart:348]: Also isn't it likely that the chomping indicator `+` is always going to cause us this kind of problems? Just asking, I'm not sure.

**Comment by jonasfj** [comment]
Wow, this starting to be a lot, I tried diving in a little bit.

My quick takeaway after seeing a few more bugs is that maybe we should be more conservative about what kind of strings we attempt to encode as literal strings.

Perhaps also as folded strings. Perhaps it's better to fallback to double quote string in the edge cases, rather than too complicated code. Certainly, we can always try to improve and not make it fallback in a future PR.

File: test/utils_test.dart (hunk 11 above)
Comment (kekavc24) [reply test/utils_test.dart]: @jonasfj The final `\n` just before the next key is the issue. 

It seems chomping indicators only control the final line break before going to the next `YamlScalar` (value in list or key in map) and not content within the scalar itself (which is  only controlled by either `>` or `|`)

After some investigation, we can exclusively use the strip chomping indicator `-` for both `ScalarStyle.FOLDED` and `ScalarStyle.LITERAL`.  It always indicates we don't want the last line break that's just before the next key in map or value in the list.

😅 The fixes I've tried coming up with for the `keep chomping indicator +` bug were all shots in the dark and just another bug creeping in. 

I'm sorry and thanks! The bug was staring right at us from the beginning. This is the line I should have removed before fixing `_tryYamlEncodeFolded`

https://github.com/dart-lang/yaml_edit/blob/f3feb50dab3b36fe5ec115e78667712473e2e882/lib/src/strings.dart#L89-L93

**Comment by kekavc24** [comment]
@jonasfj I'm sorry for the `chomping indicator` oversight on my part.

I've reduced the scope of this PR and fixed the chomping indicator issue (or rather removed the chomping indicator `+` references). If you still want me to work on [this](https://github.com/dart-lang/yaml_edit/pull/88#discussion_r1636329365) or any other issue, I am happy to.

File: lib/src/strings.dart [hunk 14]
```diff
@@ -67,142 +54,199 @@ String _yamlEncodeDoubleQuoted(String string) {
...
-          return _tryYamlEncodeLiteral(val, indentation, lineEnding);
-        }
-      }
-    }
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return '|-\n$indent${string.replaceAll('\n', lineEnding + indent)}';
+}
+
+///Encodes a flow [YamlScalar] based on the provided [YamlScalar.style].
```
Comment (jonasfj) [inline lib/src/strings.dart:182]:
```suggestion
/// Encodes a flow [YamlScalar] based on the provided [YamlScalar.style].
```

File: lib/src/strings.dart [hunk 15]
```diff
@@ -67,142 +54,199 @@ String _yamlEncodeDoubleQuoted(String string) {
...
+  final indent = ' ' * indentSize;
 
-        if (value.style == ScalarStyle.LITERAL) {
-          return _tryYamlEncodeLiteral(val, indentation, lineEnding);
-        }
-      }
-    }
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return '|-\n$indent${string.replaceAll('\n', lineEnding + indent)}';
```
Comment (jonasfj) [inline lib/src/strings.dart:179]: This is better than what we had, so let's accept it.

But I suspect this could cause issues for strings that end in empty lines or whitespace, or combination thereof.

**Review by jonasfj — APPROVED** [review]

**Comment by jonasfj** [comment]
I think we should land this as is, it's an overall improvement.

I think we should try to add more tests that challenge `_tryYamlEncodeFolded` and `_tryYamlEncodeLiteral`.
In particular with whitespace and linebreaks, in the middle, start and end.

I think we should try to just make smaller PRs that only touch either `_tryYamlEncodeFolded` or `_tryYamlEncodeLiteral`, because small improvements are easier to land.
